EIPS/eip-5994.md

```diff
@@ -0,0 +1,85 @@
+---
+eip: 5994
+title: Token Pods extensions (ERC20/ERC721)
+description: Token extension enabling external smart contract based Pods to track balances of those users who opted-in to these Pods.
+author: Anton Bukov (@k06a), Mikhail Melnik (@zumzoom)
+discussions-to: https://ethereum-magicians.org/t/eip-6000-token-pods-extensions-erc20-erc721/11782
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-11-18
+requires: 20, 721
+---
+
+## Abstract
+
+Token Pods concept introduces token extensions for the most popular token standards: ERC20, ERC721 and ERC1155, which allows holders of assets to connect their balances to Pods, which could track changes of their balances.
+
+## Motivation
+
+This Token Pods concept allows to have permissionless extensions for tokens, which can be used for incentivisations (aka farms) or delegations or any other ideas.
+
+Example implementation can be found here:
+- https://github.com/1inch/erc20-pods
+
+Examples of Pods:
+- https://github.com/1inch/farming
+- https://github.com/1inch/delegating
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+```solidity
+interface IPod {
+    function updateBalances(address from, address to, uint256 amount) external;
+}
+
+interface IERC20Pods is IERC20 {
+    function hasPod(address account, address pod) external view returns(bool);
+    function podsCount(address account) external view returns(uint256);
+    function podAt(address account, uint256 index) external view returns(address);
+    function pods(address account) external view returns(address[] memory);
+    function podBalanceOf(address pod, address account) external view returns(uint256);
+
+    function addPod(address pod) external;
+    function removePod(address pod) external;
+    function removeAllPods() external;
+}
+
+interface IERC721Pods is IERC721 {
+    function hasPod(address account, address pod) external view returns(bool);
+    function podsCount(address account) external view returns(uint256);
+    function podAt(address account, uint256 index) external view returns(address);
+    function pods(address account) external view returns(address[] memory);
+    function podBalanceOf(address pod, address account) external view returns(uint256);
+
+    function addPod(address pod) external;
+    function removePod(address pod) external;
+    function removeAllPods() external;
+}
+```
+
+### Pod specification
+
+- Pod MUST implement interface `IPod` and fit execution of `updateBalances` method in 200,000 gas and never revert.
```

Comment:
```suggestion
- Pod MUST implement interface `IPod` and fit execution of `updateBalances` method within 200,000 gas and never revert.
```

For `in`, do you mean `within`?

Could you provide rationale about how the number `200,000` was  picked??

Reply:
@xinbenlv number is relatively big for any possible logic, but small enough to have limitation 5-10 pods per account to fit into more sophisticated transaction as part of token transfer. We expect in practice users will have up to 2-3 pods with 30-50k gas consumption max for each, in average nearly 10-20k each.